### PR TITLE
UI: Detect and restore encoding and BOM in content 

### LINKED ISSUE #6716: Fix Byte Order Mark (BOM) handling in markdown display and editor.
- Gitea version (or commit ref): <= 1.8.0
- Operating system: All.
- Database (use `[x]`): All.
- Can you reproduce the bug at https://try.gitea.io:
  - [X] Yes (provide example URL) - https://try.gitea.io/emonk/Markdown-BOM-Test
  - [ ] No
  - [ ] Not relevant

## Description

When the README.md file contains a Unicode Byte Order Mark the file the first line of the file is
formatted incorrectly.  This happens for all BOMs that I have tested - UTF8, Unicode and Unicode
Big Endian.  The text of the file loads correctly in all cases, the only problem is with the BOM
being treated as file content and incorrectly displayed in both the rendered view and the editor.

This is a particular problem for developers using Visual Studio and other environments which
insert a BOM by default, and can be difficult to change without add-ons to the environment.

The various Unicode Byte Order Markers are metadata and should not be treated as file content.  At a minimum the markdown renderer should detect and discard BOMs in the content.  The markdown editor should likewise detect and remove BOMs during load and either replace them during save or save without markers.

I don't understand Go so I won't be submitting a pull request.  From 10 minutes casting about in the code it looks like [`ToUTF8WithErr`][1] and [`ToUTF8WithFallback`][2] might be a place to
start.  It looks like those are the main methods used to massage text file content into UTF8 for
both render and edit.

## Screenshots

Screenshots taken from try.gitea.io site mentioned above.

* Bad first line formatting, treats BOM as unknown non-printable character:
image
* Editor showing BOM as unknown/bad character:
image

[1]: https://github.com/go-gitea/gitea/blob/704da08fdc6bae6fdd6bf1b892ebe12afeef5eca/modules/templates/helper.go#L265
[2]: https://github.com/go-gitea/gitea/blob/704da08fdc6bae6fdd6bf1b892ebe12afeef5eca/modules/templates/helper.go#L289

### PR DESCRIPTION
**Pull request by zeripath**
Fixes #6716 

When decoding content, if the first 3 bytes match the UTF-8 BOM remove it.

When updating content through the editor, check the previous content for the encoding and BOM and reencode to  that. If we can't encode to that then default to utf8.

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by zeripath** [comment]
This could be easily backported to 1.8

**Comment by silverwind** [comment]
Tested this, can confirm it's working well. BOM is not rendered anymore and preserved if present when editing the file in the web editor.

**Review by silverwind — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Review by mrsdizzie — APPROVED** [review]

**Comment by zeripath** [comment]
Should I provide a backport?

**Comment by lafriks** [comment]
Yes please do so

File: modules/templates/helper.go [hunk 1]
```diff
@@ -267,6 +267,10 @@ func ToUTF8WithErr(content []byte) (string, error) {
 	if err != nil {
 		return "", err
 	} else if charsetLabel == "UTF-8" {
+		if len(content) > 2 && bytes.Equal(content[0:3], base.UTF8BOM) {
```
Comment (lunny) [inline modules/templates/helper.go]: How about to create a function named `RemoveUTF8BOM(content string) string`.

**Comment by lunny** [comment]
@zeripath see my comment.

File: modules/templates/helper.go (hunk 1 above)
Comment (zeripath) [reply modules/templates/helper.go]: Done

**Comment by zeripath** [comment]
I also noticed that I wasn't dealing with updating LFSed content so that's done too.

**Comment by zeripath** [comment]
(well it's dealt with in the same way we deal with it on the front end.)

File: modules/repofiles/update.go [hunk 2]
```diff
@@ -235,9 +309,26 @@ func CreateOrUpdateRepoFile(repo *models.Repository, doer *models.User, opts *Up
...
+		} else {
+			log.Error("Unknown encoding: %s", encoding)
+		}
+	}
 	var lfsMetaObject *models.LFSMetaObject
 
-	if filename2attribute2info[treePath] != nil && filename2attribute2info[treePath]["filter"] == "lfs" {
+	if setting.LFS.StartServer && filename2attribute2info[treePath] != nil && filename2attribute2info[treePath]["filter"] == "lfs" {
 		// OK so we are supposed to LFS this data!
 		oid, err := models.GenerateLFSOid(strings.NewReader(opts.Content))
```
Comment (zeripath) [inline modules/repofiles/update.go:335]:
```suggestion
		oid, err := models.GenerateLFSOid(strings.NewReader(content))
```

**Comment by zeripath** [comment]
Don't merge - I've just noticed that the LFS stuff is slightly wrong.

File: modules/repofiles/update.go [hunk 3]
```diff
@@ -235,11 +309,28 @@ func CreateOrUpdateRepoFile(repo *models.Repository, doer *models.User, opts *Up
...
+			if err != nil {
+				// Look if we can't encode back in to the original we should just stick with utf-8
+				log.Error("Error re-encoding %s (%s) as %s - will stay as UTF-8: %v", opts.TreePath, opts.FromTreePath, encoding, err)
+				result = content
+			}
+			content = result
+		} else {
+			log.Error("Unknown encoding: %s", encoding)
+		}
+	}
```
Comment (zeripath) [inline modules/repofiles/update.go:328]:
```suggestion
	}
	// Reset the opts.Content to our adjusted content to ensure that LFS gets the correct content
	opts.Content = content
```

File: modules/repofiles/update.go [hunk 4]
```diff
@@ -235,11 +309,30 @@ func CreateOrUpdateRepoFile(repo *models.Repository, doer *models.User, opts *Up
...
+	// Reset the opts.Content to our adjusted content to ensure that LFS gets the correct content
+	opts.Content = content
 	var lfsMetaObject *models.LFSMetaObject
 
-	if filename2attribute2info[treePath] != nil && filename2attribute2info[treePath]["filter"] == "lfs" {
+	if setting.LFS.StartServer && filename2attribute2info[treePath] != nil && filename2attribute2info[treePath]["filter"] == "lfs" {
 		// OK so we are supposed to LFS this data!
-		oid, err := models.GenerateLFSOid(strings.NewReader(opts.Content))
+		oid, err := models.GenerateLFSOid(strings.NewReader(content))
```
Comment (zeripath) [inline modules/repofiles/update.go]:
```suggestion
		oid, err := models.GenerateLFSOid(strings.NewReader(opts.Content))
```

**Comment by zeripath** [comment]
OK fixed!

**Pull request merged**